### PR TITLE
Port x_focus_frame

### PR DESCRIPTION
**Pull request by jubnzv**

### CODE REVIEW

File: rust_src/src/frames.rs [hunk 1]
```diff
@@ -593,4 +594,22 @@ pub fn frame_focus(frame: LispFrameLiveOrSelected) -> LispObject {
...
+/// Set the input focus to FRAME.
+/// FRAME nil means use the selected frame.  Optional argument NOACTIVATE
+/// means do not activate FRAME.
+/// If there is no window system support, this function does nothing.
+#[lisp_fn(min = "1")]
+pub fn x_focus_frame(frame: LispFrameLiveOrSelected, noactivate: LispObject) -> LispObject {
+    #[cfg(feature = "window-system")]
+    {
+        let frame_ref: LispFrameRef = frame.into();
+        let frame_ref_ptr = frame_ref.as_ptr() as *mut Lisp_Frame;
```
Comment (shaleh) [inline rust_src/src/frames.rs]: You can simply this by using `as_mut()`.

File: rust_src/src/frames.rs [hunk 2]
```diff
@@ -593,4 +594,22 @@ pub fn frame_focus(frame: LispFrameLiveOrSelected) -> LispObject {
     frame_ref.focus_frame
 }
 
+/// Set the input focus to FRAME.
+/// FRAME nil means use the selected frame.  Optional argument NOACTIVATE
+/// means do not activate FRAME.
+/// If there is no window system support, this function does nothing.
```
Comment (shaleh) [inline rust_src/src/frames.rs]: There should be an empty line with just `///` above the line 'If there is no....'.

File: rust_src/src/frames.rs [hunk 3]
```diff
@@ -8,6 +8,7 @@ use crate::{
     lisp::defsubr,
     lisp::{ExternalPtr, LispObject},
     lists::{LispConsCircularChecks, LispConsEndChecks},
+    remacs_sys::x_focus_frame as c_x_focus_frame,
```
Comment (shaleh) [inline rust_src/src/frames.rs]: No need to import this one separately. Group it with the other functions.

**Comment by agraven** [comment]
If you look at the travis results, you can see that it fails when it tries to build without a window system. The specific error it gives is `error[E0432]: unresolved import 'crate::remacs_sys::x_focus_frame'  --> src/frames.rs:11:5`.

That means you need to check at compile time whether there is a window system with `#[cfg(feature = "window-system")]`. This is the case wherever `#ifdef HAVE_WINDOW_SYSTEM` is used in C code.

**Comment by jubnzv** [comment]
Oh, it seems to work. Thank you for the thoughtful review. I just beginning in Rust and it takes some time to figure out with basics.

**Comment by shaleh** [comment]
@jubnzv this project exists in part to encourage people to learn Rust. Welcome aboard.

File: rust_src/src/frames.rs [hunk 4]
```diff
@@ -593,4 +593,23 @@ pub fn frame_focus(frame: LispFrameLiveOrSelected) -> LispObject {
     frame_ref.focus_frame
 }
 
+/// Set the input focus to FRAME.
+/// FRAME nil means use the selected frame.  Optional argument NOACTIVATE
+/// means do not activate FRAME.
+///
+/// If there is no window system support, this function does nothing.
+#[lisp_fn(min = "1")]
+pub fn x_focus_frame(_frame: LispFrameLiveOrSelected, _noactivate: LispObject) -> LispObject {
```
Comment (agraven) [inline rust_src/src/frames.rs]: You can set the type of `noactivate` to bool, that way you can pass it directly to `c_x_focus_frame`.

File: rust_src/src/frames.rs [hunk 5]
```diff
@@ -44,7 +44,7 @@ impl LispFrameRef {
             match self.vertical_scroll_bar_type() {
                 vertical_scroll_bar_type::vertical_scroll_bar_left
                 | vertical_scroll_bar_type::vertical_scroll_bar_right => {
-                    return self.config_scroll_bar_width
+                    return self.config_scroll_bar_width;
                 }
                 _ => return 0,
```
Comment (agraven) [inline rust_src/src/frames.rs]: The return statements here are redundant, you can remove them.

File: rust_src/src/frames.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (shaleh) [reply rust_src/src/frames.rs]: I am sorry @agraven, I thought I hit review earlier.

File: rust_src/src/frames.rs (hunk 5 above)
Comment (agraven) [reply rust_src/src/frames.rs]: Disregard what I just wrote, I'd actually suggest removing the added semicolon and leaving this be for now, as there's another open PR that fixes this.

File: rust_src/src/frames.rs [hunk 6]
```diff
@@ -507,13 +507,13 @@ pub fn next_frame(frame: LispFrameOrSelected, miniframe: LispObject) -> LispFram
     // a valid candidate will be returned regardless of its position.
     while passed < 2 {
         for_each_frame!(f => {
```
Comment (shaleh) [inline rust_src/src/frames.rs]: This change and the one above it do not belong and will conflict with other active PRs that are resolving the situation. Please remove them from this PR.

File: rust_src/src/frames.rs [hunk 7]
```diff
@@ -593,4 +593,22 @@ pub fn frame_focus(frame: LispFrameLiveOrSelected) -> LispObject {
...
+pub fn x_focus_frame(_frame: LispFrameLiveOrSelected, _noactivate: bool) -> LispObject {
+    #[cfg(feature = "window-system")]
+    {
+        let mut frame_ref: LispFrameRef = _frame.into();
+        unsafe {
+            c_x_focus_frame(frame_ref.as_mut(), _noactivate);
+        }
+    }
+
+    Qnil
```
[the remark this answers is not in the document]
Comment (shaleh) [reply rust_src/src/frames.rs]: If the function only and always returns `Qnil` you can define the Rust function like this:

    #[lisp_fn]
    pub fn some_function(/* args.. */) {
        // code but no returns.
    }

and the right thing will happen.

**Pull request merged**